bundles/org.openhab.binding.broadlinkthermostat/src/main/java/org/openhab/binding/broadlinkthermostat/internal/discovery/BroadlinkThermostatDiscoveryService.java

```diff
@@ -0,0 +1,174 @@
+        BLDevice[] blDevices = new BLDevice[0];
+        try {
+            InetAddress sourceAddress = getIpAddress();
+            if (sourceAddress != null) {
+                logger.debug("Using source address {} for sending out broadcast request.", sourceAddress);
+                blDevices = BLDevice.discoverDevices(sourceAddress, 0, DISCOVERY_TIMEOUT * 1000);
+            } else {
+                blDevices = BLDevice.discoverDevices(DISCOVERY_TIMEOUT * 1000);
+            }
+
+        } catch (IOException e) {
+            logger.debug("Error while trying to discover broadlinkthermostat devices: {}", e.getMessage());
+        }
+        logger.debug("Discovery service found {} broadlinkthermostat devices.", blDevices.length);
+
+        for (BLDevice dev : blDevices) {
+            logger.debug("Broadlinkthermostat device {} of type {} with Host {} and MAC {}", dev.getDeviceDescription(),
+                    Integer.toHexString(dev.getDeviceType()), dev.getHost(), dev.getMac());
+
+            ThingUID thingUID;
+            String id = dev.getHost().replaceAll("\\.", "-");
+            logger.debug("Device ID with IP address replacement: {}", id);
+            try {
+                id = getHostnameWithoutDomain(InetAddress.getByName(dev.getHost()).getHostName());
+                logger.debug("Device ID with DNS name: {}", id);
+            } catch (UnknownHostException e) {
+                logger.debug("Discovered device with IP {} does not have a DNS name, using IP as thing UID.",
+                        dev.getHost());
+            }
+
+            switch (dev.getDeviceDescription()) {
+                case "Floureon Thermostat":
+                    thingUID = new ThingUID(FLOUREON_THERMOSTAT_THING_TYPE, id);
+                    break;
+                case "Hysen Thermostat":
+                    thingUID = new ThingUID(HYSEN_THERMOSTAT_THING_TYPE, id);
+                    break;
+                default:
+                    thingUID = new ThingUID(UNKNOWN_BROADLINKTHERMOSTAT_THING_TYPE, id);
+            }
+
+            Map<String, Object> properties = new HashMap<>();
+            properties.put(BroadlinkThermostatBindingConstants.HOST, dev.getHost());
+            properties.put(Thing.PROPERTY_MAC_ADDRESS, dev.getMac().getMacString());
+            properties.put(BroadlinkThermostatBindingConstants.DESCRIPTION, dev.getDeviceDescription());
+
+            logger.debug("Property map: {}", properties);
+
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(thingUID).withProperties(properties)
+                    .withLabel(dev.getDeviceDescription() + " (" + id + ")")
+                    .withRepresentationProperty(Thing.PROPERTY_MAC_ADDRESS).build();
+
+            thingDiscovered(discoveryResult);
+
+        }
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.trace("Starting background scan for Broadlinkthermostat devices");
+        startScan();
+    }
```

Comment:
This method should return fast. You could use `scheduler.scheduleWithFixedDelay()` to let it run asynchronously.

Reply:
Changed to a scanner.